### PR TITLE
Bug: fix refresh by revision for charm hub charms

### PR DESCRIPTION
**Pull request by hmlanigan**
This is a patch for juju 2.9 to fix juju refresh by revision for charm hub charms, utilizing the `--revision` flag, based on a workaround found in the bug. The workaround was to supply a full charmurl including the new revision via the `--switch` flag. 

The real fix will involve correctly modeling both deploy and refresh actions for charm hub charms within juju and not using ResolveCharm as it is today. Today charm hub and charm store charms share ResolveCharm in both deploy and refresh code paths and it's very difficult to separate them out. In juju 3.1+, charm store charms are no longer supported by juju. As well the juju 3.1 client does not support versions prior to juju 2.9 which have no knowledge of charm hub charms.

To correctly model the refresh action, we must send both the actual current values for the application end the requested values if they have changed to charm hub.

There is a drive by to update the clean target of the Makefile, I'm tired of it not working with go 1.20.x

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [ ] ~Go unit tests, with comments saying what you're testing~
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```sh
$ make clean
```

Test charm hub charms
```sh
# revision 23 is the tip of 2.0/stable
$ juju deploy juju-qa-test --revision 22 --channel stable --series focal
# wait for refresh to complete and settle, verify revision 22
# revision 23 is the tip of 2.0/edge
$ juju refresh again --revision 23
# wait for refresh to complete and settle, verify revision 23
$ juju refresh again
# should be at revision 26, the tip of stable
```

Ensure charm store charms aren't broken here.
```sh
$ juju deploy cs:ubuntu-20 --series focal
$ juju refresh ubuntu --revision 21
$ juju refresh ubuntu
# should be at revision 24, the latest
```

```sh
$ (cd tests ; ./main.sh -v refresh)
```

## Links

https://bugs.launchpad.net/juju/+bug/1988556

JUJU-4602

### CODE REVIEW

**Review by wallyworld — DISMISSED** [review]

File: Makefile [hunk 1]
```diff
@@ -359,7 +359,8 @@ go-install: $(INSTALL_TARGETS)
 .PHONY: clean
 clean:
 ## clean: Clean the cache and test caches
-	go clean -n -r --cache --testcache $(PROJECT)/...
+	go clean -r --cache --testcache
+	go clean $(PROJECT)/...
```
Comment (jameinel) [inline Makefile:363]: clearly nobody has been using clean, given that '-n' is just print what you would do, and "-cache cannot be used with package arguments"

However, I think the actual desire is:
```
go clean --cache --testcache
go clean -r $(PROJECT)/...
```
(at least I don't think recursion makes sense for cleaning the cache)

**Review by jameinel — APPROVED** [review]
I'm quite surprised that this is all it took to have refresh to a specific revision work, but I'm happy to see it.
Presumably the issue was just in ResolveCharm?

**Comment by hmlanigan** [comment]
@jameinel I was surprised too, it's very much a patch though.

File: Makefile (hunk 1 above)
Comment (hmlanigan) [reply Makefile:363]: I use it, but put in an alias to work around it. Finally got motivated to fix it. I did wonder sometimes why things seemed off.

**Review by hpidcock — APPROVED** [review]
LGTM. Tests pass on my machine.

**Review by jameinel — APPROVED** [review]

**Pull request merged**